File: agent-driven-architecture-main/examples/typescript/src/app/contract.ts
// ── app/contract — THE ROOT: the only place that names every block (G10) ────
// The spine may not name a block and a block may not name a sibling (both G11),
// so the closed unions have exactly one legal home: here.
//
// Kotlin closes these hierarchies for itself — a sealed interface knows its own
// subtypes. TypeScript has to write the union out, and that is the whole of the
// TS/Kotlin delta in blast radius (§11.4): three extra appends in this file per
// new block. The guarantee is identical; the ceremony is not.

import type {
  AnalysisCommand,
  AnalysisEffect,
  AnalysisResult,
  AnalysisSlice,
  AnalysisView,
} from "@adr/block-analysis/register";
import { analysis } from "@adr/block-analysis/register";
import type {
  ArtifactCommand,
  ArtifactEffect,
  ArtifactResult,
  ArtifactSlice,
  ArtifactView,
} from "@adr/block-artifact/register";
import { artifact } from "@adr/block-artifact/register";
import type {
  ConsoleCommand,
  ConsoleResult,
  ConsoleSlice,
  ConsoleView,
} from "@adr/block-console/register";
import { consoleBlock } from "@adr/block-console/register";
import type {
  EscalationCommand,
  EscalationEffect,
  EscalationResult,
  EscalationSlice,
  EscalationView,
} from "@adr/block-escalation/register";
import { escalation } from "@adr/block-escalation/register";
import type { InboxCommand, InboxResult, InboxSlice, InboxView } from "@adr/block-inbox/register";
import { inbox } from "@adr/block-inbox/register";
import type {
  Priority,
  Ticket,
  TriageCommand,
  TriageEffect,
  TriageResult,
  TriageSlice,
  TriageView,
} from "@adr/block-triage/register";
import { triage } from "@adr/block-triage/register";
import type { SpineCommand } from "@adr/spine/pure/command";
import type { SpineEffect } from "@adr/spine/pure/effect";
import type { SpineSlice } from "@adr/spine/pure/spine-slice";
import { emptySpineSlice } from "@adr/spine/pure/spine-slice";
import type { SpineResult } from "@adr/spine/pure/tool-result";
import type { ViewModel } from "@adr/spine/pure/view";

// ── The three closed sets ───────────────────────────────────────────────────

export type ToolResult =
  | SpineResult
  | TriageResult
  | EscalationResult
  | ConsoleResult
  | ArtifactResult
  | AnalysisResult
  | InboxResult;

export type Command =
  | SpineCommand
  | TriageCommand
  | EscalationCommand
  | ConsoleCommand
  | ArtifactCommand
  | AnalysisCommand
  | InboxCommand;

export type Effect =
  | SpineEffect
  | TriageEffect
  | EscalationEffect
  | ArtifactEffect
  | AnalysisEffect;

/** Everything a block's arm may be handed. The spine's own two cases are folded
 *  by the spine's own arm, so they are excluded here. */
export type OkResult = Extract<ToolResult, { readonly outcome: "ok" }>;

// ── State: a PRODUCT of slices, one per block ───────────────────────────────
// State has exactly one shape, so it is a product, not a sum. Every closed set
// INSIDE it — TicketStatus, SealStatus, RunStatus, Notice — is sealed.

export interface State {
  readonly spine: SpineSlice;
  readonly triage: TriageSlice;
  readonly escalation: EscalationSlice;
  readonly console: ConsoleSlice;
  readonly artifact: ArtifactSlice;
  readonly analysis: AnalysisSlice;
  readonly inbox: InboxSlice;
}

export interface AppView extends ViewModel {
  readonly triage: TriageView;
  readonly escalation: EscalationView;
  readonly console: ConsoleView;
  readonly artifact: ArtifactView;
  readonly analysis: AnalysisView;
  readonly inbox: InboxView;
}

export interface Seed {
  readonly tickets: readonly Ticket[];
  readonly priority?: Priority;
  readonly panels?: readonly string[];
}

export function initialState(seed: Seed = { tickets: [] }): State {
  return {
    spine: emptySpineSlice,
    triage: triage.sliceOf(seed.tickets, seed.priority ?? "Normal"),
    escalation: escalation.sliceOf(seed.tickets.map((t) => t.id)),
    console: consoleBlock.sliceOf(seed.panels ?? ["escalation", "findings"]),
    artifact: artifact.emptySlice,
    analysis: analysis.emptySlice,
    inbox: inbox.emptySlice,
  };
}

File: agent-driven-architecture-main/examples/kotlin/app/src/main/kotlin/adr/app/Contract.kt
// ── app/contract — the app's closed sets (G10) ──────────────────────────────
// The ONE place in the system that may name every block. The spine may not name a
// block and a block may not name a sibling (both G11), so the product of the slices
// has exactly one legal home: the root.
//
// Note what is NOT here: the three transport unions. In Kotlin a sealed hierarchy
// closes itself — the blocks contribute their cases in adr.contract and the compiler
// already knows the set is complete. That is the whole of the TypeScript/Kotlin
// delta in §11.4: TS must write the unions out; Kotlin gets them for free.
//
// State is a PRODUCT, not a sum: this application has exactly one whole-state shape,
// so every closed set INSIDE State is sealed while State itself is a record.

package adr.app

import adr.blocks.analysis.AnalysisSlice
import adr.blocks.analysis.AnalysisView
import adr.blocks.artifact.ArtifactSlice
import adr.blocks.artifact.ArtifactView
import adr.blocks.console.ConsoleBlock
import adr.blocks.console.ConsoleSlice
import adr.blocks.console.ConsoleView
import adr.blocks.escalation.EscalationBlock
import adr.blocks.escalation.EscalationSlice
import adr.blocks.escalation.EscalationView
import adr.blocks.inbox.InboxSlice
import adr.blocks.inbox.InboxView
import adr.blocks.triage.Ticket
import adr.blocks.triage.TriageBlock
import adr.blocks.triage.TriageSlice
import adr.blocks.triage.TriageView
import adr.spine.pure.PanelId
import adr.spine.pure.SpineSlice
import adr.spine.pure.ViewModel

/**
 * Every slice defaults to its own SLICE's `empty`, so plugging a block in is ONE
 * appended field here — a seeding line is needed only for a block that starts non-empty.
 *
 * The default names the slice shape rather than the block, because a data-class default
 * is evaluated before any block exists and a block is now a CONSTRUCTED type rather than
 * a loose `object` standing by at file scope. Nothing is lost: `empty` was already the
 * idiom in inbox, artifact and analysis.
 */
data class State(
    val spine: SpineSlice = SpineSlice(),
    val triage: TriageSlice = TriageSlice(),
    val escalation: EscalationSlice = EscalationSlice(),
    val console: ConsoleSlice = ConsoleSlice(),
    val artifact: ArtifactSlice = ArtifactSlice(),
    /** The tiering rung (11): what this tier recalled, and what it published. */
    val analysis: AnalysisSlice = AnalysisSlice(),
    /** The barge-in rung (12): what was shed while busy, and what failed. */
    val inbox: InboxSlice = InboxSlice(),
)

/** The block views composed onto the spine's ViewModel root. */
data class AppView(
    val root: ViewModel,
    val triage: TriageView,
    val escalation: EscalationView,
    val console: ConsoleView,
    val artifact: ArtifactView,
    val analysis: AnalysisView,
    val inbox: InboxView,
)
